src/merkle_tree/src/storage_proof.cairo

```diff
@@ -0,0 +1,160 @@
+use core::option::OptionTrait;
+use core::array::SpanTrait;
+use core::traits::Into;
+use core::array::ArrayTrait;
+use poseidon::poseidon_hash_span;
+use debug::PrintTrait;
+
+#[derive(Drop)]
+struct BinaryNode {
+    left: felt252,
+    right: felt252,
+}
+
+#[derive(Drop, Copy)]
+struct EdgeNode {
+    child: felt252,
+    path: felt252,
+    length: u8,
+}
+
+#[derive(Drop)]
+enum TrieNode {
+    Binary: BinaryNode,
+    Edge: EdgeNode,
+}
+
+#[derive(Destruct)]
+struct ContractData {
+    class_hash: felt252,
+    nonce: felt252,
+    contract_state_hash_version: felt252,
+    storage_proof: Array<TrieNode>
+}
+
+#[derive(Destruct)]
+struct ContractStateProof {
+    class_commitment: felt252,
+    contract_proof: Array<TrieNode>,
+    contract_data: ContractData
+}
+
+/// Verify Starknet storage proof. For reference see:
+/// - ([state](https://docs.starknet.io/documentation/architecture_and_concepts/State/starknet-state/))
+/// - ([pathfinder_getproof API endpoint](https://github.com/eqlabs/pathfinder/blob/main/doc/rpc/pathfinder_rpc_api.json))
+/// - ([pathfinder storage implementation](https://github.com/eqlabs/pathfinder/blob/main/crates/merkle-tree/src/tree.rs))
+/// # Arguments
+/// * `expected_state_commitment` - state root `proof` is going to be verified against
+/// * `contract_address` - `contract_address` of the value to be verified
+/// * `storage_address` - `storage_address` of the value to be verified
+/// * `proof` - `ContractStateProof` representing storage proof
+/// # Returns
+/// * `felt252` - `value` at `storage_address` if verified, panic otherwise.
+fn verify(
+    expected_state_commitment: felt252,
+    contract_address: felt252,
+    storage_address: felt252,
+    proof: ContractStateProof
+) -> felt252 {
+    let contract_data = proof.contract_data;
+
+    let (contract_root_hash, storage_value) = traverse(
+        storage_address, contract_data.storage_proof
+    );
+
+    let contract_state_hash = pedersen(
+        pedersen(pedersen(contract_data.class_hash, contract_root_hash), contract_data.nonce),
+        contract_data.contract_state_hash_version
+    );
```

Comment:
This could/should be updated to:
```suggestion
    let contract_state_hash = PedersenTrait::new(contract_data.class_hash)
        .update_with(contract_root_hash)
        .update_with(contract_data.nonce)
        .update_with(contract_data.contract_state_hash_version)
        .finalize();
```
You'll have to adapt import to add:
```rust
use pedersen::{PedersenTrait, HashState};
use hash::{HashStateTrait, HashStateExTrait};
```

Reply:
find it more readable to see smthing updated with x, y, z than fn(fn(fn(,...))) but that's just my 2c